### PR TITLE
Max height

### PR DESCRIPTION
**Pull request by chris-luangrath**
In order for this to work, also needs edits from the max-height branch in BYU-PCCL/holodeck (the python/json side)

HolodeckAgent.h - GetMaxHeight function and MaxHeight float
I feel like I should change how it’s implemented. I just kinda put it there, and I feel like it’s doing its own thing.

HolodeckGameMode.h - Updated SpawnAgent function

SpawnAgentCommand.cpp - Changed IsMainAgent = NumberParams[6] to [7] 
(I’m kinda not sure where I changed that made it like this tbh, but I edited the code to make it work. Should I find a way to make it 7 again, or does it not matter?)
float MaxHeight = NP[6] (Same concern as [6] to [7])
Updated SpawnAgent call to include MaxHeight
MaxHeight is now a var in SpawnAgent

Uav.cpp
SetMaxHeight(GetMaxHeight())
SetMaxHeight function
I feel like I *could* consolidate SetMaxHeight into MaxHeightCeiling, but I wouldn't want to add another line of code to run every tick when it only needs to run once
MaxHeightCeiling function teleports the Uav to the maxHeight-1 if it's above the maxHeight

### CODE REVIEW

File: Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp [hunk 1]
```diff
@@ -5,7 +5,7 @@
 void USpawnAgentCommand::Execute() {
 
 	UE_LOG(LogHolodeck, Log, TEXT("SpawnAgentCommand::Execute spawning agent"));
-	verifyf(StringParams.size() == 2 && NumberParams.size() == 7, TEXT("%s: Bad Arguments"), *FString(__func__));
+	verifyf(StringParams.size() == 2 && NumberParams.size() <= 8, TEXT("%s: Bad Arguments"), *FString(__func__)); // NumberParams.size() used to be == 7
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp]: shouldn't it be `==8` now then?

File: Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp [hunk 2]
```diff
@@ -26,18 +26,20 @@ void USpawnAgentCommand::Execute() {
 	// Note that we have to re-order the parameters since FRotator takes pitch, roll, yaw
 	// but the coordinates from the Python side com in roll, pitch, yaw order
 	FRotator Rotation = FRotator(NumberParams[4], NumberParams[3], NumberParams[5]);
-	bool IsMainAgent = (bool) NumberParams[6];
+	float MaxHeight = NumberParams[6];
+	bool IsMainAgent = (bool) NumberParams[7]; // used to be 6, now is 7
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp]: You don't need the comment here

File: Source/Holodeck/Agents/Private/Uav.cpp [hunk 3]
```diff
@@ -22,6 +22,8 @@ AUav::AUav() {
 void AUav::InitializeAgent() {
 	Super::InitializeAgent();
 	RootMesh = Cast<UStaticMeshComponent>(RootComponent);
+
+	SetMaxHeight(GetMaxHeight());
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/Agents/Private/Uav.cpp]: Why isn't SetMaxHeight being called here and not in the spawn agent command?

File: Source/Holodeck/Agents/Public/Uav.h [hunk 4]
```diff
@@ -75,4 +77,6 @@ class HOLODECK_API AUav : public AHolodeckAgent
 	  */
 	void InitializePIDControllers();
 
+	float maxHeight;
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/Agents/Public/Uav.h]: You have a maxHeight variable for the uav and a MaxHeight for its parent class HolodeckAgent. You only need the parent class one.

File: Source/Holodeck/HolodeckCore/Public/HolodeckAgent.h [hunk 5]
```diff
@@ -125,6 +125,11 @@ class HOLODECK_API AHolodeckAgent : public AHolodeckAgentInterface {
 	TMap<FString, UHolodeckSensor*> SensorMap;
 	AHolodeckPawnControllerInterface* HolodeckController;
 
+
+	virtual float GetMaxHeight();
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/HolodeckCore/Public/HolodeckAgent.h]: Doesn't need to be virtual since the child classes aren't overriding it.

**Review by MaxRobinsonTheGreat — CHANGES_REQUESTED** [review]

**Comment by MaxRobinsonTheGreat** [comment]
Also, we're merging into develop, not master.

**Review by chris-luangrath — COMMENTED** [review]
Everything should be fixed now.

File: Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp [hunk 6]
```diff
@@ -26,18 +26,20 @@ void USpawnAgentCommand::Execute() {
 	// Note that we have to re-order the parameters since FRotator takes pitch, roll, yaw
 	// but the coordinates from the Python side com in roll, pitch, yaw order
 	FRotator Rotation = FRotator(NumberParams[4], NumberParams[3], NumberParams[5]);
-	bool IsMainAgent = (bool) NumberParams[6];
+	float MaxHeight = NumberParams[6];
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp]: Convert to cm by multiplying by 100 here.

File: Source/Holodeck/HolodeckCore/Private/HolodeckAgent.cpp [hunk 7]
```diff
@@ -122,3 +122,14 @@ bool AHolodeckAgent::InitializeController() {
 		return true;
 	}
 }
+
+void AHolodeckAgent::MaxHeightCeiling() {
+	if (GetActorLocation().Z > MaxHeight - 1) {
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/HolodeckCore/Private/HolodeckAgent.cpp]: Why -1?

File: Source/Holodeck/HolodeckCore/Public/HolodeckGameMode.h [hunk 8]
```diff
@@ -53,7 +53,7 @@ class AHolodeckGameMode : public AGameMode
 
 	// These functions allow the Holodeck to do things which cannot normally be done from pure c++ code
 	UFUNCTION(BlueprintImplementableEvent)
-	AHolodeckAgent* SpawnAgent(const FString& Type, const FVector& Location, const FRotator& Rotation, const FString& Name, bool IsMainAgent);
+	AHolodeckAgent* SpawnAgent(const FString& Type, const FVector& Location, const FRotator& Rotation, const FString& Name, bool IsMainAgent, float MaxHeight);
```
Comment (MaxRobinsonTheGreat) [inline Source/Holodeck/HolodeckCore/Public/HolodeckGameMode.h]: Shouldn't there be a change to the holodeckgamemode blueprint file in this pr?
Also, can you remove the changes to the example level map? You can do that with git

**Review by MaxRobinsonTheGreat — DISMISSED** [review]

**Comment by vinhowe** [comment]
@MaxRobinsonTheGreat Did all of your comments get resolved?

File: Source/Holodeck/Agents/Private/Uav.cpp [hunk 9]
```diff
@@ -35,6 +35,12 @@ void AUav::ApplyForces() {
 	float YawTorqueToApply = FMath::Clamp(GetYawTorqueToApply(), -UAV_MAX_YAW_RATE, UAV_MAX_YAW_RATE);
 	float ThrustToApply = FMath::Clamp(GetThrustToApply(), -UAV_MAX_FORCE, UAV_MAX_FORCE);
 
+	if (GetActorLocation().Z >= MaxHeight) {
+		if (ThrustToApply >= 0) {
+			ThrustToApply = 0;
+		}
+	}
+
```
Comment (vinhowe) [inline Source/Holodeck/Agents/Private/Uav.cpp]: Did we verify that this makes a difference?

File: Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp [hunk 10]
```diff
@@ -26,7 +26,8 @@ void USpawnAgentCommand::Execute() {
 	// Note that we have to re-order the parameters since FRotator takes pitch, roll, yaw
 	// but the coordinates from the Python side com in roll, pitch, yaw order
 	FRotator Rotation = FRotator(NumberParams[4], NumberParams[3], NumberParams[5]);
-	bool IsMainAgent = (bool) NumberParams[6];
+	float MaxHeight = (NumberParams[6] * 100);// -11.8283691; // *100 to convert to centimeters, and for whatever reason, subtracting 11.8... makes it more accurate. Could be height of
```
Comment (vinhowe) [inline Source/Holodeck/ClientCommands/Private/SpawnAgentCommand.cpp]: If we're not subtracting here, can we remove the comment?

**Review by vinhowe — CHANGES_REQUESTED** [review]

**Review by vinhowe — APPROVED** [review]
Looks good to me

**Review by daniekpo — APPROVED** [review]
LGTM! Good work!

**Pull request merged**